osf/models/collection.py

```diff
@@ -0,0 +1,153 @@
+    is_bookmark_collection = models.BooleanField(default=False, db_index=True)
+    is_deleted = models.BooleanField(default=False, db_index=True)
+
+    def __unicode__(self):
+        return '{self.title!r}, with guid {self._id!r}'.format(self=self)
+
+    @property
+    def url(self):
+        return '/{}/'.format(self._id)
+
+    def get_absolute_url(self):
+        return self.absolute_api_v2_url
+
+    @property
+    def absolute_api_v2_url(self):
+        return api_v2_url('/collections{}'.format(self.url))
+
+    @property
+    def linked_nodes_self_url(self):
+        return self.absolute_api_v2_url + 'relationships/linked_nodes/'
+
+    @property
+    def linked_registrations_self_url(self):
+        return self.absolute_api_v2_url + 'relationships/linked_registrations/'
+
+    @property
+    def linked_nodes_related_url(self):
+        return self.absolute_api_v2_url + 'linked_nodes/'
+
+    @property
+    def linked_registrations_related_url(self):
+        return self.absolute_api_v2_url + 'linked_registrations/'
+
+    def save(self, *args, **kwargs):
+        first_save = not self.id
+        if self.is_bookmark_collection:
+            if self.title != 'Bookmarks':
+                # Bookmark collections are always named 'Bookmarks'
+                self.title = 'Bookmarks'
+            if not self.pk and self.creator.collection_set.filter(is_bookmark_collection=True, is_deleted=False).exists():
+                raise IntegrityError('Each user cannot have more than one Bookmark collection.')
+        ret = super(Collection, self).save(*args, **kwargs)
+        if first_save:
+            # Set defaults for M2M
+            self.collected_types = ContentType.objects.filter(app_label='osf', model__in=['abstractnode', 'collection'])
+        return ret
+
+    def collect_object(self, obj, collector, collected_type=None, status=None):
+        """ Adds object to collection, creates CollectedGuidMetadata reference
+            Performs type / metadata validation. User permissions checked in view.
+
+        :param GuidMixin obj: Object to collect. Must be of a ContentType specified in collected_types
+        :param OSFUser collector: User doing the collecting.
+        :param str collected_type: Metadata "type" of submission, validated against collected_type_choices
+        :param str status: Metadata "status" of submission, validated against status_choices
+        :return: CollectedGuidMetadata object or raise exception
+        """
+        collected_type = collected_type or ''
+        status = status or ''
+
+        if self.collected_type_choices and collected_type not in self.collected_type_choices:
+            raise ValidationError('Unacceptable "type".')
+
+        if self.status_choices and status not in self.status_choices:
+            raise ValidationError('Unacceptable "status".')
+
+        if not any([isinstance(obj, t.model_class()) for t in self.collected_types.all()]):
```

Comment:
I think this is a little clearer and would raise immediately instead of hitting all types
```python
for t in self.collected_types.all():
    if isinstance(obj, t.model_class()):
        raise ValidationError('Unacceptable ContentType.')
```

Reply:
Except that only errors if it's an acceptable type.